File: framework/languages/java/samples/java/TicTacToe.SessionGateway/Server/Session/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/server/session/sessions/handlers/JoinMatchSessionPacketHandler.java
package systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.handlers;

import java.util.concurrent.CompletionStage;
import systems.zlink.contracts.messaging.Message;
import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionPacketHandler;
import systems.zlink.framework.streams.ZLinkStreamHeader;
import systems.zlink.samples.tictactoe.sessiongateway.shared.configuration.SampleNames;

public final class JoinMatchSessionPacketHandler
    implements ZLinkSessionPacketHandler<ZLinkSessionContext> {
    private final ZLinkClient channels;
    private final PlayNotificationRelay relay;

    public JoinMatchSessionPacketHandler(
        ZLinkClient channels,
        PlayNotificationRelay relay) {
        this.channels = channels;
        this.relay = relay;
    }

    @Override
    public String packetName() {
        return "JoinMatchReq";
    }

    @Override
    public CompletionStage<Void> handleAsync(
        ZLinkSessionContext context,
        ZLinkStreamHeader header,
        Message payload) {
        ZLinkSessionActor actor = CreateMatchSessionPacketHandler.requireSingleBoundActor(context);
        return channels.requestToChannel(
                SampleNames.PlayChannel,
                payload.toUtf8String().trim() + "|" + actor.actorId())
            .packetName("JoinMatchReq")
            .submitAsync(String.class)
            .thenCompose(response -> relay.deliverAsync(response)
                .thenCompose(ignored -> context.client()
                    .reply(relay.reply(response))
                    .submitAsync()));
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe.SessionGateway/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/sessiongateway/server/session/sessions/handlers/JoinMatchSessionPacketHandler.kt
package systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.handlers

import java.util.concurrent.CompletionStage
import systems.zlink.contracts.messaging.Message
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionPacketHandler
import systems.zlink.framework.streams.ZLinkStreamHeader
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.shared.configuration.SampleNames

class JoinMatchSessionPacketHandler(
    private val channels: ZLinkClient,
) : ZLinkSessionPacketHandler<ZLinkSessionContext> {
    override fun packetName(): String = "JoinMatchReq"

    override fun handleAsync(
        context: ZLinkSessionContext,
        header: ZLinkStreamHeader,
        payload: Message,
    ): CompletionStage<Void> {
        val actor = CreateMatchSessionPacketHandler.requireSingleBoundActor(context)
        return channels.requestToChannel(
            SampleNames.PlayChannel,
            "${payload.toUtf8String().trim()}|${actor.actorId()}",
        )
            .packetName("JoinMatchReq")
            .submitAsync(String::class.java)
            .thenCompose { reply ->
                context.client()
                    .send(reply)
                    .packetName(SampleNames.TurnChangedPacket)
                    .submitAsync()
                    .thenCompose {
                        context.client()
                            .reply(reply)
                            .submitAsync()
                    }
            }
    }
}
